libzeth/snarks/groth16/groth16_api_handler.tcc

```diff
@@ -72,17 +72,21 @@ void groth16_api_handler<ppT>::extended_proof_to_proto(
     const extended_proof<ppT, groth16_api_handler<ppT>::snark> &ext_proof,
     zeth_proto::ExtendedProof *message)
 {
-    libsnark::r1cs_gg_ppzksnark_proof<ppT> proof_obj = ext_proof.get_proof();
+    // Note that the protobuf format exports -b (`minus_b`) to support a small
+    // optimization in on-chain verification code.
```

Comment:
Shall this one be rephrased to reflect that the proposed changes are mostly for simplicity matters?
Maybe something like: "// Note that the protobuf format exports -b (`minus_b`) to simplify (and slightly optimize) the on-chain verification code."

Reply:
Additionally, could you make sure to document this on top of the function declaration in the `hpp` file using Doxygen's syntax to make sure this shows up in the generated docs? (same for `_from_proto` to explain that `minus_b` is negated back to `b`)

Reply:
Here https://github.com/clearmatics/zeth/blob/negate-groth16-b/libzeth/snarks/groth16/groth16_api_handler.hpp#L29-L35

Reply:
BTW, why not doing the negation on the python client in the `to_contract_parameters`? Wouldn't that save the need to modify the cpp API?

Reply:
> BTW, why not doing the negation on the python client in the proof_to_contract_parameters?

Yes, I did consider that.  I can see the point - we limit the scope of this change to the encoding made by the client for the verification code on the contract.  The main reason I didn't do that here is because the client would then need to know which curve is being used, and the value of `q` for that curve.  Currently it does not have that information and just uses a single scheme that converts field elements and curve points to evm words.

We could do that.  It would require some `CURVE` config in the client, and the curve constants (or maybe the client could query the server to get `q`, or to actually perform the negation)  but it's definitely possible.

Reply:
If we take this apraoach, it might make more sense to just do a one-time negation of some elements of the verification key and upload that to the contract at deployment.  We rely on the prover server being available at deployment, so we could add an RPC method to get the curve parameters and the client could just use those for a one-shot transform.  Proofs can then remain unchanged with `b`.

Reply:
(This would involve negating `vk.delta` and one of `vk.alpha` or `vk.beta`.  The contract would then contain a hard-coded `$[-1]_2$` instead of `$[1]_2$` in the verifier, or it could be passed a constructor parameter).

Reply:
One other thought: if we add an RPC method it could be `GetProcessedVerificaitonKey` which returns the VK with extra data: `-beta`, `-delta` and `-1` in G2.

Reply:
Exact. I think this would be quite a nice thing to have actually (some curve config on the client), here are the pros/cons I can think of (please add any that comes to mind)
Pros:
- By having access to some "remarkable params" (field characteristics etc), we can not only do the negation above, but also support some tests on the received arguments (type/"range" check received elements if needed - this btw needs to be systematically done on the contract to avoid any potential attacks leveraging the mismatch in the length of EVM words and field elements -> that's bad because this advocates to define partial functions (and an API with partial functions is.. well.. annoying to say the least), but that's a requirement for some minimal type safety on the contract)
- We keep consistent APIs and limit changes to the smallest scope possible

Cons:
- Needs to track the curve config on the client (which exacerbates the "configuration logistic" where several params need to be duplicated to cpp code, python code and solidity code -> this may be fixed later on by building some tooling/scripts (like a lightweight `configure` script to generate each components' config for us to lessen this issue - but out of scope now))

To this point btw, we can surely avoid such troubles by extending the server's API to have a "fetchServerConfig" endpoint that fetches the server configuration/protocol params from the server and configures the client as per the fetched data. That way the curve/pairing group name (i.e. `ALT_BN128`, `BLS12_377` etc) can be part of the received message, and upon reception of the `init request` response the client can instantiate `CurveParams` by selecting the right local curve configuration and proceed (i.e. as soon as the connection is established with the server/as soon as the client is started -> it fetches the server config).
This seems to be the right tradeoff. It keeps the client generic (works with any pairing group), while keeping things fairly simple. I can't think about any meaningful use-case where the client keeps processing proofs coming from several servers configured using different curves (so adding the "curve" to the `extended_proof` object - which would solve against that - seems completely overkilled for instance).

Reply:
Oh, I just read your comment:
> We rely on the prover server being available at deployment, so we could add an RPC method to get the curve parameters and the client could just use those for a one-shot transform. Proofs can then remain unchanged with b.

For some reason it didn't show up when I wrote my comment (maybe I forgot to refresh the page before commenting, not sure). Anyway, we seems to align with the point I made at the end of my previous comment. That looks like a potential solution.

Reply:
Explained `minus_b` in the doc comments for this class.  Ideally it would be on the protobuf type declaration, but that may not be captured by doc generation so added on the class.